i18n/iconv.rs

```diff
@@ -0,0 +1,471 @@
+        ),
+        Encodings::ASCII => ascii::from_ucs4(input.as_slice(), omit_invalid, supress_error),
+    };
+
+    let exit_code = input_exit_code.max(output_exit_code);
+
+    (output, exit_code)
+}
+fn charmap_conversion(
+    from: &Charmap,
+    to: &Charmap,
+    input: &[u8],
+    omit_invalid: bool,
+    suppress_error: bool,
+) -> (Vec<u8>, u32) {
+    let mut output = Vec::new();
+    let mut error_count = 0;
+
+    let mut i = 0;
+    while i < input.len() {
+        let mut found = false;
+        for (_, entry) in &from.entries {
+            if input[i..].starts_with(&entry.encoding) {
+                if let Some(to_entry) = to
+                    .entries
+                    .values()
+                    .find(|e| e.symbolic_name == entry.symbolic_name)
+                {
+                    output.extend_from_slice(&to_entry.encoding);
+                    i += entry.encoding.len();
+                    found = true;
+                    break;
+                }
+            }
+        }
+
+        if !found {
+            if !suppress_error {
+                eprintln!("Error: Invalid or unmapped character at position {}", i);
+            }
+            error_count += 1;
+            if omit_invalid {
+                i += 1;
+            } else {
+                output.push(input[i]);
+                i += 1;
+            }
+        }
+    }
+
+    let exit_code = if error_count > 0 { 1 } else { 0 };
+    (output, exit_code)
+}
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let args = Args::parse();
+
+    setlocale(LocaleCategory::LcAll, "");
+    textdomain(PROJECT_NAME)?;
+    bind_textdomain_codeset(PROJECT_NAME, "UTF-8")?;
+
+    if args.list_codesets {
+        list_encodings();
+        exit(0);
+    }
+
+    let from_codeset = args.from_codeset.unwrap_or_else(|| {
+        env::var("LANG")
+            .ok()
+            .and_then(|lang| lang.split('.').nth(1).map(String::from))
+            .unwrap_or_else(|| {
+                eprintln!("Error: Could not find a codeset from your locale");
+                exit(1);
+            })
+    });
+
+    let to_codeset = args.to_codeset.unwrap_or_else(|| {
+        env::var("LANG")
+            .ok()
+            .and_then(|lang| lang.split('.').nth(1).map(String::from))
+            .unwrap_or_else(|| {
+                eprintln!("Error: Could not find a codeset from your locale");
+                exit(1);
+            })
+    });
+
+    let from_codeset = parse_codeset(&from_codeset)?;
+    let to_codeset = parse_codeset(&to_codeset)?;
+
+    let inputs: Vec<Box<dyn Read>> = match args.files {
+        Some(files) => files
+            .into_iter()
+            .map(|file| plib::io::input_stream(&file, true))
+            .collect::<Result<Vec<_>, _>>()?,
+        None => vec![Box::new(io::stdin().lock())],
+    };
+
+    for mut input in inputs {
+        let mut inp_buf = Vec::new();
+        input.read_to_end(&mut inp_buf)?;
```

Comment:
What level of difficulty is a streaming implementation?

An all-in-memory implementation limits users to documents to a maximum of 1/2 available memory + swap.

Reply:
well, a hurdle for me was getting the boundary of the last character of a specific encoding.
i'll take a look into it's solution today

Reply:
actually, i can probably work "directly" with the Read trait object and extracting just what i need throughout the conversion

Reply:
hmm..doesn't seem simple ..i'll try to figure out a way